### DISCUSSION TITLE
Server Actions Fail to Import Members from Packages in Monorepo

### DISCUSSION THREAD

**Discussion by john-prutton** [vercel/next.js#58371] [Help]
### Summary

# The issue
I have a turborepo set up with one nextjs app and a database package. 

I can successfully use the database packing in server components and api routes, but when I use it in a server action there is an error. For some reason it can't seem to import anything from the database package.  

## Repo
[`liquiFruit/like-it-wear-it`@`feat/add-db`](https://github.com/liquiFruit/like-it-wear-it/tree/feat/add-db)

## Successfully using the package in RSC
Here I have an RSC page that can import and use the `getUsers` function from the database package.

![image](https://github.com/vercel/next.js/assets/19376890/afce34b1-68e7-4496-84f4-177f4a4393e0)

## Error using the package in a server action
Here I have a client component with a `onClick` event handler on a button. This event handler calls a server action `tryCreateUser`, which in turn calls a function from the database package.

![image](https://github.com/vercel/next.js/assets/19376890/e5ca546d-ad19-4a06-a99e-65ccb4f9abfc)

## A (terrible) solution
Upon moving `"use server"` into the database package, the client component can directly import the function `createUser` from the database package and successfully invoke it.

![image](https://github.com/vercel/next.js/assets/19376890/992601f5-b290-4038-9451-9e341e05c49f)

## Conclusion
This is terrible for many reasons: 
1. The database package should be completely decoupled from nextjs  as I will consume it from multiple apps (reactjs, nestjs, etc)
2.  I can then only export async functions from the database package, whereas I need to export objects too (`db`, `schema`, etc)

### Additional information

```js
next.js: 14.0.1
```

### Example

https://github.com/liquiFruit/like-it-wear-it/tree/feat/add-db

**Comment by Pascal-Lohscheidt** [comment]
So I still have that issue on 14.2.23. (I am using Prisma though) and its unable to find the prisma db client

And even though similar things are mentioned in other threads  and they praise setting esmExternals: false. It did not work for me.

However I realised it only happens for some server actions and only to those which where on a route where the there was no layout and the page.tsx was client side or did not fetch anything from the database and did not import the database client. 

I assume it might be because of what is written [here](https://nextjs.org/docs/14/app/building-your-application/data-fetching/server-actions-and-mutations#authentication-and-authorization) 
> Server Actions inherit the [runtime](https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes) from the page or layout they are used on.

So for me making sure that the page.tsx or layout.tsx (from which I am calling the server-action) is using the database helped.

just placing that here if someone is stumbling upon that problem themself
